### DISCUSSION TITLE
pydantic 2 - disable BaseModel validation

### DISCUSSION THREAD

**Discussion by pynkpanther** [pydantic/pydantic#7128] [Question]
- We use pydantic for all our FastAPI microservices.
- We used to disable model validation for services with high throughput by overriding the `__init__ ` method of our models inherited from `BaseModel`:

```
class MyModel(BaseModel):
  a: int

if disable_validation:
  model.__init__ = lambda self, **data: None
```

- this however does not seem to work anymore, payload of requests to the service is always validated.
- despite the pydantic 2 ``BaseModel.__init__` still looks like overrding it would disable validation:

```
    def __init__(__pydantic_self__, **data: Any) -> None:  # type: ignore
        """Create a new model by parsing and validating input data from keyword arguments.

        Raises [`ValidationError`][pydantic_core.ValidationError] if the input data cannot be
        validated to form a valid model.

        `__init__` uses `__pydantic_self__` instead of the more common `self` for the first arg to
        allow `self` as a field name.
        """
        # `__tracebackhide__` tells pytest and some other tools to omit this function from tracebacks
        __tracebackhide__ = True
        __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
```

- i also tried overriding `MyModel.model_json_schema`, writing my own `SchemaValidator` class with own `validate_python` / `isinstance_python` / `validate_json` methods but this doesnt seem to disable validation for FastAPI requests either.

can someone point me to the where I should look?

Regards
Klaus

**Comment by smivv** [comment]
I would ask in a more generic way: How to disable validation for specific `BaseModel`?
Would be nice to have just one flag in the `ConfigDict` that disables it by default.

**Comment by paulchaum** [comment]
I know this is an old thread, but I had the same need and managed to fix it by making middleware that bypasses the Pydantic validation. So I paste it here in case somebody else needs it:
```
import mcp.types as mt
from fastmcp.server.middleware import CallNext, Middleware, MiddlewareContext
from fastmcp.tools.base import ToolResult
from fastmcp import FastMCP


class _BypassOutputSchemaValidationMiddleware(Middleware):
    """Bypass MCP SDK's lowlevel ``outputSchema`` validation on tool calls.

    Each tool advertises a Pydantic-derived ``outputSchema`` so MCP clients
    know the expected response shape. The lowlevel MCP SDK normally runs
    ``jsonschema.validate(structured_content, outputSchema)`` on the way out
    (see ``mcp/server/lowlevel/server.py`` -- ``Output validation error``).
    BoondManager API responses occasionally drift from the OpenAPI-derived
    schemas (extra/missing fields), and our service layer already handles
    that leniently via :func:`boondmanager_tools.libs.boond_manager.validation.validate`
    (``model_construct`` fallback). Letting the lowlevel SDK fail here would
    surface those drifts to the client as opaque ``Output validation error``
    messages, defeating the lenient policy.

    FastMCP's ``ToolResult.to_mcp_result`` returns a :class:`CallToolResult`
    directly when ``meta`` is not ``None``, which bypasses the lowlevel
    validation step. We use the same trick FastMCP itself uses in
    ``ResponseLimitingMiddleware``: set ``meta`` to an empty dict (or
    preserve any existing meta) on the result before it is serialized.
    """

    async def on_call_tool(
        self,
        context: MiddlewareContext[mt.CallToolRequestParams],
        call_next: CallNext[mt.CallToolRequestParams, ToolResult],
    ) -> ToolResult:
        result = await call_next(context)
        if isinstance(result, ToolResult) and result.meta is None:
            result.meta = {}
        return result


mcp = FastMCP("My MCP Server")

# Add middleware to bypass output schema validation
mcp.add_middleware(_BypassOutputSchemaValidationMiddleware())
```

**Comment by gistrec** [comment]
The reason your `__init__` override no longer works isn't that `__init__` got harder to monkey-patch - it's that **FastAPI doesn't call `__init__` to build the model from a request**. In v2, `BaseModel.__init__` just delegates to `__pydantic_validator__.validate_python` (a `pydantic_core.SchemaValidator` written in Rust, wired up at class creation), and FastAPI bypasses `__init__` entirely: it builds its own `TypeAdapter` for the body parameter at route-registration time and calls the validator directly. So patching `__init__` after the class exists has no effect on the request path.

The supported "skip validation" primitive in v2 is `model_construct`:

```python
class MyModel(BaseModel):
    a: int

m = MyModel.model_construct(a="not-an-int")  # no validation, no coercion
```

For FastAPI specifically, the only approach I've seen hold up under load is to take the body parsing away from FastAPI for those routes:

```python
from fastapi import Request
import orjson

@app.post("/ingest")
async def ingest(request: Request):
    payload = MyModel.model_construct(**orjson.loads(await request.body()))
    ...
```

That skips both Pydantic validation and FastAPI's own JSON-decoding/validation layer, which is usually a bigger win than the validation itself. If you want a middle ground - keep the typed signature but cheaper validation - use a `TypedDict` plus `TypeAdapter`; `pydantic_core` validates `TypedDict` significantly faster than a `BaseModel` because it skips the `__dict__`/setattr machinery.

I'd avoid trying to swap `__pydantic_validator__` on the class globally. It can work in isolated scripts, but FastAPI caches the `TypeAdapter` per parameter at startup, so the swap has to happen before the route is registered, and you lose coercion (`"1"` → `1`) which often matters more than full validation does.

Bottom line: `model_construct` is the right tool, and for FastAPI you have to apply it on the request side rather than expecting the model to opt out.

**Comment by s-chudmunge** [comment]
The reason your `__init__` override no longer works isn't that `__init__` is harder to patch — it's that **FastAPI doesn't call `__init__` to build the model from a request body**. In Pydantic v2, `BaseModel.__init__` just delegates to `__pydantic_validator__.validate_python` (a Rust-backed `pydantic_core.SchemaValidator` wired up at class-creation time). FastAPI builds its own `TypeAdapter` for each body parameter at route-registration time and calls that validator directly — bypassing `__init__` entirely. Patching `__init__` after the class is defined has no effect on the request path.

## The supported primitive: `model_construct`

The official "skip validation" entry point in v2 is `model_construct`:

```python
class MyModel(BaseModel):
    a: int

m = MyModel.model_construct(a="not-an-int")  # no coercion, no validation
```

## For FastAPI routes specifically

You have to take body parsing away from FastAPI for those routes:

```python
from fastapi import Request
import orjson
from pydantic import BaseModel

class MyModel(BaseModel):
    a: int

@app.post("/ingest")
async def ingest(request: Request):
    payload = MyModel.model_construct(**orjson.loads(await request.body()))
    ...
```

This skips both Pydantic validation and FastAPI's own JSON-decoding/validation layer, which is typically a larger performance win than validation alone.

## A faster middle ground

If you want a typed signature with cheaper validation rather than none at all, use a `TypedDict` plus `TypeAdapter`. `pydantic_core` validates `TypedDict` significantly faster than `BaseModel` because it skips `__dict__`/setattr machinery:

```python
from typing import TypedDict
from pydantic import TypeAdapter

class MyData(TypedDict):
    a: int

adapter = TypeAdapter(MyData)

@app.post("/ingest")
async def ingest(request: Request):
    payload = adapter.validate_python(orjson.loads(await request.body()))
```

## Caveats

- Avoid swapping `__pydantic_validator__` on the class globally. FastAPI caches the `TypeAdapter` per parameter **at startup**, so the swap must happen before the route is registered, and you lose coercion (`"1"` → `1`) which often matters more than full validation overhead.
- `model_construct` skips both validation *and* default factory calls unless you pass `_fields_set`; keep that in mind for models with computed defaults.
